### PR TITLE
add RecMLM

### PR DESCRIPTION
**Pull request by junieberry**

### CODE REVIEW

File: finetune_data/process.py [hunk 1]
```diff
@@ -47,10 +48,14 @@ def extract_meta_data(path):
...
             title = line["title"]
 
+            meta_data[asin] = attr_dict
             if len(title) != 0:
                 attr_dict["title"] = title
                 attr_dict["brand"] = brand
                 attr_dict["category"] = category
+                if args.desc:
+                    description = line["description"]
+                    attr_dict["description"] = description
```
Comment (sudokim) [inline finetune_data/process.py:58]: Description이 없는 경우에는 어떻게 되나요?

Comment (sudokim) [inline finetune_data/process.py:58]: Description이 없는 경우에는 어떻게 되나요?

File: recmlm/models.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        self.max_attr_num = max_attr_num
+        self.max_attr_length = max_attr_length
+
+
+class RecMLM(pl.LightningModule):
+
+    def __init__(
+            self,
+            model: LongformerForMaskedLM,
+            learning_rate: float = 1e-5,):
```
Comment (sudokim) [inline recmlm/models.py:33]: Formatting...

File: recmlm/models.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    def training_step(self, batch, batch_idx):
+        input_ids = batch["input_ids"]
+        attention_mask = batch["attention_mask"]
+        global_attention_mask = batch["global_attention_mask"]
+        label = batch["label"]
+        outputs = self(input_ids, attention_mask, global_attention_mask, label)
+        loss = outputs.loss
+
+        self.log("train_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
```
Comment (sudokim) [inline recmlm/models.py:58]: `train/loss`로 접어주면 wandb에서 train끼리 묶어서 보여준답니다.

File: recmlm/models.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        outputs = self(input_ids, attention_mask, global_attention_mask, label)
+        loss = outputs.loss
+        self.log("test_loss", loss, on_step=True, on_epoch=True, prog_bar=True, logger=True)
+        return loss
+
+    def configure_optimizers(self):
+
+        optimizer = torch.optim.AdamW(self.parameters(), lr=self.learning_rate)
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=3, verbose=True)
+        return {'optimizer': optimizer, 'lr_scheduler': scheduler, 'monitor': 'val_loss'}
```
Comment (sudokim) [inline recmlm/models.py:85]: 마지막 빈 줄 부탁드려요

File: recmlm/dataset.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+        else:
+            _, self.valid_dataset, self.test_dataset = self.get_dataset()
+            torch.save(self.valid_dataset, "valid_dataset.pt")
+            torch.save(self.test_dataset, "test_dataset.pt")
+
+    def get_dataset(self, train=False):
+        train_histories = list(list(self.user2train.values())[:-1])
+        valid_histories = list(list(self.user2train.values()))
+        test_histories = [self.user2train[user] + self.user2val[user] for user in self.user2val.keys()]
```
Comment (sudokim) [inline recmlm/dataset.py:42]: 이거는 안 바꾸기로 한건가요? (그냥 궁금)

File: longformer_mlm.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        val_check_interval=args.val_check_interval,
+        accumulate_grad_batches=args.accumulate_grad_batches,
+        callbacks=callbacks,
+        logger=loggers,
+        gradient_clip_val=args.gradient_clip_val,
+    )
+
+    trainer.fit(module, datamodule=datamodule)
+    trainer.test(ckpt_path="best", datamodule=datamodule)
+    trainer.save_checkpoint(path_output / Path("model.ckpt"))
```
Comment (sudokim) [inline longformer_mlm.py:112]: Config 저장도 해주시죠

File: recmlm/dataset.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        for history in tqdm(histories, ncols=120, desc=f"Creating dataset"):
+            history = [self.item_meta_dict[self.id2item[item_id]] for item_id in history]
+            tokenized_history = self.tokenizer(history)
+            tokenized_history['label'] = tokenized_history['input_ids'].copy()
+            masked_input_ids = []
+            for idx, (input_id, token_id) in enumerate(
+                    zip(tokenized_history['input_ids'], tokenized_history['token_type_ids'])):
+                is_mask = False
+                if token_id == 0:  # bos token
+                    masked_input_ids.append(input_id)
```
Comment (sudokim) [inline recmlm/dataset.py:69]: `is_mask`가 if문 아래에 들어가야 하는 것 아닌가요?

File: recmlm/dataset.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+                elif token_id == 1:  # item token
+                    is_mask = False
+                    masked_input_ids.append(input_id)
+                elif token_id == 2 and tokenized_history['token_type_ids'][idx - 1] == 1:
```
Comment (sudokim) [inline recmlm/dataset.py:73]: 이 분기점의 의미는 무엇인가요?

File: recmlm/dataset.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                    # mask with mlm_ratio
+                    is_mask = True if torch.rand(1) < self.mlm_ratio else 0
+                    masked_input_ids.append(self.tokenizer.mask_token_id if is_mask else input_id)
+                elif token_id == 2:
+                    masked_input_ids.append(self.tokenizer.mask_token_id if is_mask else input_id)
```
Comment (sudokim) [inline recmlm/dataset.py:78]: 이렇게 masking하게 되면 attribute 하나를 통째로 가리는게 아니라 사실상 랜덤 토큰을 가리는 것 아닌가요?

**Review by sudokim — CHANGES_REQUESTED** [review]
확인 부탁드려요

File: recmlm/dataset.py (hunk 7 above)
Comment (junieberry) [reply recmlm/dataset.py:69]: 헐..

Comment (junieberry) [reply recmlm/dataset.py:69]: 진짜 헐랭이다

Comment (junieberry) [reply recmlm/dataset.py:69]: 고마워요.. 어쩐지 loss가 작더라

File: recmlm/dataset.py (hunk 8 above)
Comment (junieberry) [reply recmlm/dataset.py:73]: token type id가 0 111 2222222 이렇게 구성되어 있을 때
0은 bos, 1은 key, 2는 value입니다.
이전 token type id가 1이고 현재 token type id가 2일때 key에서 value로 진입했다는 뜻이기에
0.15의 확률로 연속되게 등장하는 2들을 마스킹합니다...

구리구리코드 죄송합니다

File: recmlm/dataset.py (hunk 9 above)
Comment (junieberry) [reply recmlm/dataset.py]: 수정했습니다ㅠㅠ

File: longformer_mlm.py (hunk 6 above)
Comment (junieberry) [reply longformer_mlm.py:145]: ㅎㅎ굿

File: recmlm/dataset.py (hunk 5 above)
Comment (junieberry) [reply recmlm/dataset.py:36]: 네... 저 안 바꾸고 싶어요

File: recmlm/models.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        for name, param in longformer_state_dict.items():
+            if name not in recformer_state_dict:
+                continue
+            else:
+                try:
+                    if not recformer_state_dict[name].size() == param.size():
+                        recformer_state_dict[name].copy_(param)
+                except:
+                    continue
+        recformer.load_state_dict(recformer_state_dict, strict=False)
```
Comment (sudokim) [inline recmlm/models.py:119]: 이렇게 한번 해보시고 알려주실래요?

```suggestion
        recformer.longformer.load_state_dict(self.model.state_dict(), strict=False)
```

**Review by sudokim — CHANGES_REQUESTED** [review]
